### ISSUE TITLE
assert (var := nullable-expr) makes var still nullable when accessed from another module

### ISSUE THREAD

**Issue by rdong8** [microsoft/pyright#9566]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

If you have a question about typing or a behavior that you’re seeing in Pyright (as opposed to a bug report or enhancement request), consider posting to the [Pyright discussion forum](https://github.com/microsoft/pyright/discussions).

**Describe the bug**
A clear and concise description of the behavior you are seeing and the expected behavior along with steps to reproduce it.

Combining an assert not null with the walrus operator at the global scope in one module causes the assigned variable to fail to be recognized as non-null from another module. It works when the assignment and assertion are separated into different lines.

**Code or Screenshots**
If possible, provide a minimal, self-contained code sample (surrounded by triple back ticks) to demonstrate the issue. The code should define or import all referenced symbols.

main.py:

```python
import a

a: int = a.a # Type "int | None" is not assignable to declared type "int"
```

a.py:

```python
assert (a := {'a': 1}.get('a'))
```

If your code relies on symbols that are imported from a third-party library, include the associated import statements and specify which versions of those libraries you have installed.

**VS Code extension or command-line**
Are you running pyright as a VS Code extension, a language server in another editor, integrated into Pylance, or the command-line tool? Which version?

Extension

**Comment by erictraut** [comment]
Pyright is working as designed here, so this isn't a bug.

A symbol whose type is narrowing via a conditional expression (including an `assert` statement) applies only within a local execution context (a function or module). Local type narrowing does not apply across module boundaries. This behavior is not specific to the walrus operator or `assert` statements. You'll see the same behavior in the example below.

```python
from random import random

a = 1 if random() > 0.5 else None
if a is None:
    raise ValueError()

reveal_type(a)  # int
```

I'll also note that using an assert in this manner is pretty dangerous because assert statements are completely skipped at runtime if the Python code is run using the `-O` option. You should not place program logic in an assert statement. They should be used only to document assumptions about values.

**Comment by rdong8** [comment]
Sorry I'm still unconvinced. When I do this:

```python
assert (x := <expr that returns nullable int>)
```

No matter if I am in the local scope or in an external module, `x` has type `int` no? As in there isn't type narrowing, but either `x` exists and is an int, or it's unbound because someone caught the assertion error exception. But the statement "in all control paths where `x` is initialized, it is an int" is true.
